### PR TITLE
Fix/protocol stats totals

### PR DESCRIPTION
**Pull request by BetsyBraddock**
## Motivation and Context

This is branched from feat/Stake Data. Please review feat/Stake Data first.

This PR fixes a bug with the protocol stats totals calculation for AMM LP fees and AMM staker fees.

## Checklist

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] the pull request title describes what this PR does (not a vague title like `Update index.md`)
- [ ] the pull request targets the *default* branch of the repository (`development`)
- [ ] the code follows the established code style of the repository
  - `npm run lint` passes
- [ ] the code build without any errors
  - `npm run codegen` passes
  - `npm run build` passes
- [ ] tests are added for the changes I made (if any source code was modified)
- [ ] documentation added or updated
- [ ] I have run the project locally and verified that there are no errors

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Fixes

fixes #[issue number], fixes #[issue number]

### CODE REVIEW

File: scripts/sql/views/core_protocol_fees.sql [hunk 1]
```diff
@@ -0,0 +1,50 @@
+CREATE VIEW core_protocol_fees as
```
Comment (omerzam) [inline scripts/sql/views/core_protocol_fees.sql]: what is this and why do i see this in 2 different PRs?

File: scripts/sql/views/tokens_current.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+ CREATE VIEW tokens_current as
```
Comment (omerzam) [inline scripts/sql/views/tokens_current.sql]: what is this and why do i see it in 2 different PRs?

**Review by omerzam — COMMENTED** [review]
there seems to be a lot in common with the 2 recent PRs, why?
it also looks like both PRs do a lot more than they were suppose to do
let's try to keep prs isolated to their scope

**Comment by BetsyBraddock** [comment]
> there seems to be a lot in common with the 2 recent PRs, why? it also looks like both PRs do a lot more than they were suppose to do let's try to keep prs isolated to their scope

Sorry, it's branched from Stake Data PR (it's in the PR description). Once Stake Data is merged to dev, the changes won't show in this PR anymore. But yes, I will try to keep the scope smaller.

File: scripts/sql/views/core_protocol_fees.sql (hunk 1 above)
Comment (BetsyBraddock) [reply scripts/sql/views/core_protocol_fees.sql]: This and the other one are the views that I'm using for the data warehouse. I've been re-running this sql to recreate the views when I switch over the data warehouse subgraph and I needed somewhere to store the sql query.

Eventually I want to just create the views automatically in each subgraph by running the sql query when the subgraph is created (like with the database triggers)

**Comment by BetsyBraddock** [comment]
@omerzam is this ok to merge?

File: src/LiquidityPoolConverter.ts [hunk 3]
```diff
@@ -204,12 +205,12 @@ class IConversionEvent {
 }
 
 function handleConversion(event: IConversionEvent): void {
-  const fromAmount = decimalize(event.fromAmount, event.fromToken)
-  const toAmount = decimalize(event.toAmount, event.toToken)
-  const conversionFee = decimalize(event.conversionFee, event.toToken)
-  const protocolFee = decimalize(event.protocolFee, event.toToken)
+  const toToken = Token.load(event.toToken.toHexString())
+  const fromAmount = decimal.fromBigInt(event.fromAmount, DEFAULT_DECIMALS)
+  const toAmount = decimal.fromBigInt(event.toAmount, DEFAULT_DECIMALS)
```
Comment (omerzam) [inline src/LiquidityPoolConverter.ts]: shouldn't we use toToken decimals here?

File: src/LiquidityPoolConverter.ts [hunk 4]
```diff
@@ -204,12 +205,12 @@ class IConversionEvent {
 }
 
 function handleConversion(event: IConversionEvent): void {
-  const fromAmount = decimalize(event.fromAmount, event.fromToken)
-  const toAmount = decimalize(event.toAmount, event.toToken)
-  const conversionFee = decimalize(event.conversionFee, event.toToken)
-  const protocolFee = decimalize(event.protocolFee, event.toToken)
+  const toToken = Token.load(event.toToken.toHexString())
+  const fromAmount = decimal.fromBigInt(event.fromAmount, DEFAULT_DECIMALS)
```
Comment (omerzam) [inline src/LiquidityPoolConverter.ts]: shouldn't this be fromToken decimals here?

File: src/LiquidityPoolConverter.ts [hunk 5]
```diff
@@ -204,12 +205,12 @@ class IConversionEvent {
...
+  const toAmount = decimal.fromBigInt(event.toAmount, DEFAULT_DECIMALS)
+  const conversionFee = decimal.fromBigInt(event.conversionFee, DEFAULT_DECIMALS)
+  const protocolFee = decimal.fromBigInt(event.protocolFee, DEFAULT_DECIMALS)
```
Comment (omerzam) [inline src/LiquidityPoolConverter.ts]: which token is the protocolFee?
if it is in SOV, shouldn't it be then the SOV decimals (even if it is 18)

File: src/LiquidityPoolConverter.ts (hunk 4 above)
Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts]: Sort of, but I'm using default decimals because:
1. We don't currently have the subgraph set up to handle non-18 decimal tokens
2. The fromToken Token entity isn't loaded already here. It gets loaded at least 3 times throughout handleConversion, which is very slow, so I don't want to load it again. Eventually I want to refactor this so it's only loaded once and then passed to each successive function as a param and only saved at the end

File: src/LiquidityPoolConverter.ts (hunk 5 above)
Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts]: protocolFee is in the toToken. I'm trying not to load entities that aren't strictly needed because it slows down syncing, especially with events like Conversion that are very frequent.

I want to refactor this so that toToken and fromToken are loaded only once (see above comment)

File: src/LiquidityPoolConverter.ts (hunk 3 above)
Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts]: Same as comment about fromToken. I know toToken Token entity is actually loaded here, but I wanted to use default decimals in both or neither to keep them consistent

File: src/LiquidityPoolConverter.ts (hunk 4 above)
Comment (omerzam) [reply src/LiquidityPoolConverter.ts]: > Sort of, but I'm using default decimals because:
> 
>     1. We don't currently have the subgraph set up to handle non-18 decimal tokens
> 
>     2. The fromToken Token entity isn't loaded already here. It gets loaded at least 3 times throughout handleConversion, which is very slow, so I don't want to load it again. Eventually I want to refactor this so it's only loaded once and then passed to each successive function as a param and only saved at the end

1. what does that mean? what is not set up for using non-18 decimal tokens?
2. let's add this refactor as a task on Jira and add it to the backlog

File: src/LiquidityPoolConverter.ts (hunk 3 above)
Comment (omerzam) [reply src/LiquidityPoolConverter.ts]: how is it consistent if the decimals are different and potentially the calculation is wrong?

File: src/LiquidityPoolConverter.ts (hunk 4 above)
Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts]: Oh actually I think you're right - it is set up to handle non-18 tokens on the AMM, it's just not set up for non-BTC pairs yet.

I've added the refactor task here: https://sovryn.atlassian.net/browse/SOV-685

And I'll change this to use correct decimals

File: src/LiquidityPoolConverter.ts (hunk 5 above)
Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts]: I've changed this to use token decimals for everything, am marking as resolved

File: src/LiquidityPoolConverter.ts [hunk 6]
```diff
@@ -206,11 +206,14 @@ class IConversionEvent {
 
 function handleConversion(event: IConversionEvent): void {
   const toToken = Token.load(event.toToken.toHexString())
-  const fromAmount = decimal.fromBigInt(event.fromAmount, DEFAULT_DECIMALS)
-  const toAmount = decimal.fromBigInt(event.toAmount, DEFAULT_DECIMALS)
-  const conversionFee = decimal.fromBigInt(event.conversionFee, DEFAULT_DECIMALS)
-  const protocolFee = decimal.fromBigInt(event.protocolFee, DEFAULT_DECIMALS)
+  const fromToken = Token.load(event.fromToken.toHexString())
+  const toDecimals = toToken !== null ? toToken.decimals : DEFAULT_DECIMALS
```
Comment (omerzam) [inline src/LiquidityPoolConverter.ts:210]: toToken can be null? in which cases?

File: src/LiquidityPoolConverter.ts [hunk 7]
```diff
@@ -206,11 +206,14 @@ class IConversionEvent {
...
+  const fromAmount = decimal.fromBigInt(event.fromAmount, fromToken !== null ? fromToken.decimals : DEFAULT_DECIMALS)
```
Comment (omerzam) [inline src/LiquidityPoolConverter.ts:211]: let's make the logic consistent between toDecimals and fromDecimals
in which cases is fromToken === null?
also shouldn't this be fromToken != null and not fromToken !== null?

Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts:211]: it's never null, but that code won't compile without the nullability check

Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts:211]: In the newer version of assemblyscript, nullability should be checked by triple equals: https://thegraph.com/docs/en/release-notes/assemblyscript-migration-guide/

In the version we use now it doesn't matter, but this will make it easier to upgrade

Comment (omerzam) [reply src/LiquidityPoolConverter.ts:211]: I think for objects this nullability check is invalid with our current version

Comment (BetsyBraddock) [reply src/LiquidityPoolConverter.ts:211]: We're using both in the subgraph already, and both seem to be working (a triple equals that definitely works here: https://github.com/DistributedCollective/Sovryn-subgraph/blob/b6d1b76f073382ba10323d8d121a5536b304d4fa/src/utils/Candlesticks.ts#L91)

**Review by omerzam — APPROVED** [review]
